### ISSUE TITLE
the problem of calculating grad 

### ISSUE THREAD

**Issue by hexunshi** [pytorch/pytorch#3827]
I register two hook and calculate the backward of  loss function for batch training
```python
def backward_hook(module, grad_input, grad_output):
    module.register_buffer('backward_hook_input',grad_input)
    module.register_buffer('backward_hook_output',grad_output)
    return None
def forward_hook(module,input,output):
    module.register_buffer('forward_hook_input',input)
    module.register_buffer('forward_hook_output',output)
    return None
```
```python
optimizer = torch.optim.SGD(net.parameters(), lr=0.5)
loss_func = torch.nn.MSELoss() 
forward_hook1 = net.hidden.register_forward_hook(hook=forward_hook)
backward_hook1 = net.hidden.register_backward_hook(backward_hook)
prediction = net(x[0:15])
loss = loss_func(prediction, y[0:15])
optimizer.zero_grad()
loss.backward()
forward_hook1.remove()
backward_hook1.remove()
```
calculate the following two values
```python
torch.matmul(net.hidden.forward_hook_input[0].transpose(0,1),net.hidden.backward_hook_output[0])
net.hidden.weight.grad.transpose(0,1)
```
Why are these two numbers equal?
I think the grad of weight is the sum of the grad of each data,like this
```python
for i in range(0,15):
    sum += torch.matmul(net.hidden.forward_hook_input[0][i].transpose(0,1),net.hidden.backward_hook_output[0][i])
```

**Comment by zou3519** [comment]
It's a bit difficult to tell what's going on here, especially because I don't know what the size of your data is and what `net` is. Could you provide more code or more mathematical context to what exactly you're calculating?
